### PR TITLE
Spike approach to refactoring validator

### PR DESCRIPTION
**Pull request by iwootten**
### PR Context

Adds multiple validator classes for validating schemas/questionnaires/questions and answers to break up the current validator class. Each class operates on a schema element (dict), collects errors/context and has a number of static error messages to prevent string comparisons throughout tests. Some existing error messages have had their content modified to remove contextual information from the message and place it in a separate context dict.

The new answer validator adds unit tests for it to replace our previous integration style schemas and the existing questionnaire validator has had a few of the existing tests migrated across to use unit tests which create instances of classes rather than the entire app. Visibility of methods have been modified in order to support unit tests.

### CODE REVIEW

**Comment by iwootten** [comment]
[Spike approach to refactoring validator](https://trello.com/c/Yxp8vwjg/3717-spike-approach-to-refactoring-validator)

**Comment by ajmaddaford** [comment]
Some thoughts (probably best discussed as a team in a tech session):
- The answer validator contains a lot of validation for different answer types. We should consider breaking it down into validators for these types.
- There is some routing validation in the answer validator. We should only be validating the schema element relevant to the validator. Routing validation should either be done in a block validator or potentially as a separate routing validator.
- The JSON schema validation should be separated out to run independently.
- We should pre-process the schema to collect relevant properties e.g. list names or block ids, rather than combine it with the validation.
- There are some inconsistencies e.g. how errors are collected, but I'm guessing that's because not all the code has been updated.
- What happened to the QuestionnaireSchema module you were working on?

**Comment by iwootten** [comment]
@ajmaddaford Agree with all of these. I'd also like to move the use of static answer validator methods out of the questionnaire validator. The schema validator is something I would have liked to do but felt the changes to unit tests were the thing I wanted most and was really pushing the boundaries of the word "spike" by the end.

For the QuestionnaireSchema stuff, I still have it but I'm not sure what to do with it? This now also includes the graph visualisation I linked to in slack before. We could create another repo for it and include it as per eq-translations. I wonder if this will make updating runner frustrating though?

File: app/validation/answer_validator.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+        self.list_names = list_names
+        self.block_ids = block_ids
+
+    @cached_property
+    def options(self):
+        return self.answer.get("options", [])
+
+    def validate(self):
+        self.validate_duplicate_options()
+        self._validate_answer_actions()
```
[the remark this answers is not in the document]
Comment (iwootten) [reply app/validation/answer_validator.py:30]: Agree, I've basically attempted to migrate the tests that already exist and see where coverage highlights issues. There's many other protected methods that should really be public. 

Unfortunately validate() does a lot of the setup of objects for passing to tests e.g. _validate_blocks, _validate_group_routing_rules. This means that we end up calling the main validate in tests, which in turn calls everything, increasing code coverage 🙄. We should try not to do this in new unit tests and instead write smaller tests.

If we leave these as protected, we can at least see where we aren't doing explicit tests.

**Pull request closed without merging**